### ISSUE TITLE
Application hangs until debugger paused and continued

### ISSUE THREAD

**Issue by kinex** [Dart-Code/Dart-Code#1714]
This happens occasionally while I am debugging my app:
1. The app is not responding anymore (or at least some async operation seems to hang, the app UI may partially work). The app has not stopped to any breakpoint or exception.
2. After I hit the debugger pause button and then continue button, the app works normally again.

**Comment by DanTup** [comment]
Could you try reproducing while capturing a log?

https://dartcode.org/docs/logging/#capture-logs-command

Run **Dart: Capture Logs**, tick the **Debugger** option (and if it's a Flutter app, also the **Flutter Run** option), then reproduce the issue. Then click **Stop Logging** (note: the logging notification may have collapsed into the bell icon in the status bar) and provide the log.

Thanks!

**Comment by kinex** [comment]
I will try this when I see the issue next time. Issue seems to require some specific debugging/IDE state as usually I don't see this for several weeks, but for example today I was able to reproduce it several times.

**Comment by DanTup** [comment]
Please let me know if you're still able to repro this and grab a log and I'll re-open this. I don't think there's anything I can progress without more info.

**Comment by kinex** [comment]
Logs sent by email.

**Comment by DanTup** [comment]
Can you send a screenshot of what the VS Code debug side bar and toolbar looks like when this happens? The log suggests that it's paused at a breakpoint (in `src/reminders/reminder_notification_manager.dart`), but it doesn't sound like the UI is reflecting that correctly.

Also please check if there are any errors in the debug console (Help -> Toggle Developer Tools, then Console tab).

Are you able to reproduce the issue in a small project you can share?

**Comment by kinex** [comment]
More info sent by email. I noticed code execution continues also if I just add or remove a breakpoint. Implementing a sample project may be impossible, because I have no idea what should I try to do there. I have seen this issue happening in multiple places in my app, it is not only these current source files I am now debugging.

**Comment by DanTup** [comment]
I misread the log earlier - although it stopped at a breakpoint, it looks like you resumed it before this happened. After you resumed, Flutter was sending Frame events for a short while then seems to have just stopped (around the time of `_documentChangeToItemChange`).

I don't think this problem is in the VS Code extension, as the log otherwise looks fine. Are you able to easily test the same thing with Android Studio or from the terminal with `flutter run` to confirm it still happens there? If so, we could open an issue in flutter/flutter where hopefully someone may have ideas for debugging (I'm much less familiar with the internals of Flutter).

**Comment by kinex** [comment]
Yes, there was a breakpoint after the "deadlock" position. When the issue happens, the breakpoint is never hit (it is hit only after I wake up the debugger again with the pause/resume button or add a new breakpoint anywhere). 

I found this issue months ago, but now I found UI steps how I can reproduce the issue quite easily (at least at the moment). It may still require 5-10 attempts before it occurs. I also unchecked "All Exceptions" and "Uncaught Exceptions" and added only one breakpoint. Issue still occurs in VS Code. Also worth of notice is that when it hangs (with the same UI steps), it seems to always hang to the exactly same position (based on what I see in the Debug Console log).

I tested this quickly also with Android Studio using a breakpoint at the same source file and source line, using the same UI steps. I could not reproduce the issue yet (~30 attempts). 

Not sure where the bug is, but from my understanding this kind of bug cannot be in my code. If it was a deadlock in my app, just adding a breakpoint to IDE (or clicking the debugger Pause+Resume buttons should not "fix" the deadlock). This is actually the worst consequence of this issue. You may be investigating a deadlock issue in your app for days (as I have already done), but actually there is a tools issue.

Let me know if you have other ideas how to help finding the issue. I will test with Android Studio again after I have installed the latest updates to it.

**Comment by DanTup** [comment]
> Not sure where the bug is, but from my understanding this kind of bug cannot be in my code.

I would be inclined to think that too, but I don't know enough about the internals of Flutter (for example if there are native plugins, how they interact with Dart threads).

Something else to try - get into this state where it seems to have paused, and then run the **Open Observatory** command from VS Code - this'll open a web app with a debugger. Go to the Debugger section and see what it says about the isolates (for ex. if they're paused, running, idle, whatever). Maybe we can figure out where it is from there (it shows better information on the status of the running isolates than we do in VS Code).

**Comment by kinex** [comment]
I tried that **Open Observatory**, but as soon as I launch it the VS Code debugger resumes automatically and runs to the next breakpoint. So I cannot see details from the paused state.

But now I found the exact code line where it seems to hang. It usually hangs to `File.create` in a code block similar to this:

```dart
var file = File(fullPath);
print("create ${file.path}..");
file = await file.create(recursive: true);
print("created ${file.path}");
```
I use that code block whenever I need to create a new file or open an existing file so it is called quite often.

When it hangs, I only see the "create" in log. And as soon as I modify breakpoints (or click the pause+resume buttons), the "created" log is printed (for the same file). Maybe in about 9 of 10 cases it hanged to this same line. In one case it hanged to some other location.

This would be much simpler if it hanged to `File.create` permanently (would be an issue in `File.create`). But the fact that e.g. modifying breakpoints "fixes" the issue makes this really confusing. 

For some reason I still cannot reproduce this in Android Studio. I tried it over 100 times with no luck. Despite of this fact I wouldn't bet anymore the issue is in Dart Code either.

**Comment by DanTup** [comment]
@bkonyi any idea where to go with this? The VM seems to sometimes just unexpectedly stop doing anything (eg. Flutter frames stop coming, app seems unresponsive). Various actions like clicking Pause/Resume, modifying breakpoints, or even just opening the Observatory web app seem to kick it back to life.

I hoped using Observatory we could see what the isolate is doing, but since opening it in the browser causes it to continue makes that tricky.

**Comment by DanTup** [comment]
I wonder if this is related? It also seems to be a hang on an `await`, possibly in a callback from Firebase?

https://github.com/flutter/flutter/issues/36179

**Comment by kinex** [comment]
I am using Firestore (that other case uses Firebase Realtime Database).

My current understanding is that this issue happens only while debugging the app. My app is already in open beta testing and no related issues reported, although it is true that this issue may be difficult to detect in app UI.

**Comment by DanTup** [comment]
Ah, I'd forgotten about that issue.

@kinex do you think you'd be able to make a repro you could share for @bkonyi to investigate?

**Comment by kinex** [comment]
Surprisingly I was able to create a repro project very easily. This doesn't of course mean that you can reproduce it in your development environment and test device, but please try it.

Steps:
1. Create a new Flutter project (Flutter 1.7.8+hotfix4).
2. Edit main.dart as follows:

```dart
import 'package:path_provider/path_provider.dart';
.
.
class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  int _fileId = 0;
  Future test() async {
    final dir = await getApplicationDocumentsDirectory();
    ++_fileId;
    final filePath = dir.path + "/file$_fileId.txt";
    final file = File(filePath);
    for (int i = 0; i < 1000; i++) {
      print("$i Creating ${file.path}...");
      await file.create(recursive: true);
      print("...$i created ${file.path}");
      await Future.delayed(Duration(milliseconds: 1));

      await file.writeAsString("$i", flush: true);

      await Future.delayed(Duration(milliseconds: 1));
    }
  }

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
    // add a breakpoint here
    test();
  }
  @override
  Widget build(BuildContext context) {
    // original code here
  }  
}

```
3. Add a breakpoint to the `test()` call.
4. Start debugging in VS Code (NOT Android Studio).
5. Hit the + button in device to launch the `test()` function.
6. Breakpoint is hit, click F5 in VS Code to continue.
7. Follow Debug Console to verify that `test()` is still running. At some point log messages stop appearing there.
8. Wait few seconds and continue from step 5 until log messages stop appearing. In my tests ~10 iterations were needed.
9. After log messages stop appearing, it is the error state. Try adding a new breakpoint to resume debugger.

**Comment by kinex** [comment]
Is the fix mentioned in dart-lang/sdk#37312 included in Flutter 1.7.8+hotfix.4?

**Comment by DanTup** [comment]
> Is the fix mentioned in dart-lang/sdk#37312 included in Flutter 1.7.8+hotfix.4?

Good question! I'm having trouble matching up the Dart version in Flutter with the Dart SDK, so I can't be sure.

I tried your repro steps above, but I can't reproduce the issue (using the iOS Simulator at least):

```
flutter: 998 Creating /Users/dantup/Library/Developer/CoreSimulator/Devices/31B1C998-21CC-4A32-ADB1-0C84568FA24A/data/Containers/Data/Application/431EA605-9B8F-4E04-A7B2-A245681A1497/Documents/file4.txt...
flutter: ...998 created /Users/dantup/Library/Developer/CoreSimulator/Devices/31B1C998-21CC-4A32-ADB1-0C84568FA24A/data/Containers/Data/Application/431EA605-9B8F-4E04-A7B2-A245681A1497/Documents/file4.txt
flutter: 999 Creating /Users/dantup/Library/Developer/CoreSimulator/Devices/31B1C998-21CC-4A32-ADB1-0C84568FA24A/data/Containers/Data/Application/431EA605-9B8F-4E04-A7B2-A245681A1497/Documents/file4.txt...
flutter: ...999 created /Users/dantup/Library/Developer/CoreSimulator/Devices/31B1C998-21CC-4A32-ADB1-0C84568FA24A/data/Containers/Data/Application/431EA605-9B8F-4E04-A7B2-A245681A1497/Documents/file4.txt
```

I'm using the same version of Flutter:

```
Flutter 1.7.8+hotfix.4 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 20e59316b8 (6 weeks ago) • 2019-07-18 20:04:33 -0700
Engine • revision fee001c93f
Tools • Dart 2.4.0
```

I created a new project using **Flutter: New Project**, added `path_provider` to pubspec and fetched packages, when replaced the state class (minus the build method) with your code. Anything else you had to do?

**Comment by kinex** [comment]
Those should be the correct steps and all versions are same. I used this command line to create the project:
```
flutter create --project-name testapp testapp
```
But I tested this using device (Galaxy S7 / Android 8) and my development machine has Windows 10. I can reproduce the issue every time, it requires about 10 button clicks.

It would be essential to know if we already have the fix, because it looks very similar issue.

**Comment by DanTup** [comment]
Oh, I thought the loop was to trigger it, I only clicked 4 or 5 times.

FWIW, I found the Dart SDK included in Flutter stable, and it's https://github.com/dart-lang/sdk/tree/7340a569caac6431d8698dc3788579b57ffcf0c6 which does *not* include the fix above.

So, it's very possible switching to a newer branch like `beta` or `dev` will resolve the issue if it's the same. If you still see it there, let me know and I'll do some more testing here to see if I can repro (I do have a Windows 10 machine and some Android devices I can test with too).

**Comment by kinex** [comment]
I switched to `dev` channel, same issue still.

[√] Flutter (Channel dev, v1.9.5, on Microsoft Windows [Version 10.0.18362.295])
    • Flutter version 1.9.5 at C:\tools\fluttersdk\flutter
    • Framework revision 365f577c70 (5 days ago), 2019-08-23 14:08:04 -0700
    • Engine revision 72283b6419
    • Dart version 2.5.0

[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at C:\Program Files (x86)\Android\android-sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = C:\Program Files (x86)\Android\android-sdk
    • ANDROID_SDK_ROOT = C:\Program Files (x86)\Android\android-sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)
    • All Android licenses accepted.

[√] Android Studio (version 3.5)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 38.2.3
    • Dart plugin version 191.8423
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)

[√] VS Code (version 1.37.1)
    • VS Code at C:\Users\kinex\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.3.0

[√] Connected device (1 available)
    • SM G930F • ad0c16439214b951e3 • android-arm64 • Android 8.0.0 (API 26)

**Comment by maxim-saplin** [comment]
I noticed similar app freezes in debug mode (both in VM running tests and on Android emulator) when the app creates multiple isolates and waits for them to complete - to unfreeze the app I go to Debug/Stack Trace in VS Code and do pause/resume on every one of running isolates and after that the app continues execution.

**Comment by DanTup** [comment]
@bkonyi did you happen to get chance to look at the repro above? Does it makes sense to move this to an SDK issue (it occurs to me it's easily overlooked here as it's probably not something that can be handled here). Thanks!

**Comment by kinex** [comment]
I had already forgotten this issue, but I haven't seen this happening in my apps anymore for a long time.

**Comment by bkonyi** [comment]
> @bkonyi did you happen to get chance to look at the repro above? Does it makes sense to move this to an SDK issue (it occurs to me it's easily overlooked here as it's probably not something that can be handled here). Thanks!

No, I totally forgot about this issue. Can you file an issue in the SDK and assign it to me so I don't lose track?

**Comment by DanTup** [comment]
I tried with the repro above and ran it 20 times and was also unable to reproduce it, so perhaps it has been fixed somewhere along the way.

I'll close this, but should you see it again please file an issue in the SDK and ping me/@bkonyi. Thanks!

**Comment by jjoelson** [comment]
@DanTup I still run into this issue relatively frequently. I finally decided to search for it and came across this issue. The symptoms described above are _very_ familiar, so I don't think this was fixed.

As far as I can tell, the mere existence of breakpoints seems to occasionally cause asynchronous operations to never complete. Usually removing the breakpoints causes the futures to resume execution.

Is there something in particular I could post here that would help narrow down the issue?
